internal/controllers/cluster/cluster_controller_phases.go

```diff
@@ -143,132 +141,149 @@ func (r *Reconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.C
+		r.recorder.Eventf(cluster, corev1.EventTypeNormal, "InfrastructureReady", "Cluster %s InfrastructureReady is now True", cluster.Name)
+	}
+	cluster.Status.InfrastructureReady = true
+
 	return ctrl.Result{}, nil
 }
 
 // reconcileControlPlane reconciles the Spec.ControlPlaneRef object on a Cluster.
-func (r *Reconciler) reconcileControlPlane(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+func (r *Reconciler) reconcileControlPlane(ctx context.Context, s *scope) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
+	cluster := s.cluster
 
 	if cluster.Spec.ControlPlaneRef == nil {
 		return ctrl.Result{}, nil
 	}
 
 	// Call generic external reconciler.
-	controlPlaneReconcileResult, err := r.reconcileExternal(ctx, cluster, cluster.Spec.ControlPlaneRef)
+	obj, err := r.reconcileExternal(ctx, cluster, cluster.Spec.ControlPlaneRef)
 	if err != nil {
-		return ctrl.Result{}, err
-	}
-	// Return early if we need to requeue.
-	if controlPlaneReconcileResult.RequeueAfter > 0 {
-		return ctrl.Result{RequeueAfter: controlPlaneReconcileResult.RequeueAfter}, nil
-	}
-	// If the external object is paused, return without any further processing.
-	if controlPlaneReconcileResult.Paused {
-		return ctrl.Result{}, nil
-	}
-	controlPlaneConfig := controlPlaneReconcileResult.Result
+		if apierrors.IsNotFound(err) {
+			s.controlPlaneIsNotFound = true
 
-	// There's no need to go any further if the control plane resource is marked for deletion.
-	if !controlPlaneConfig.GetDeletionTimestamp().IsZero() {
-		return ctrl.Result{}, nil
+			if !cluster.DeletionTimestamp.IsZero() {
+				// Tolerate control plane not found when the cluster is being deleted.
+				return ctrl.Result{}, nil
+			}
+
+			if cluster.Status.ControlPlaneReady {
+				// Control plane went missing after the cluster was up and running
+				return ctrl.Result{}, reconcile.TerminalError(errors.Errorf("%s has been deleted after being ready", cluster.Spec.ControlPlaneRef.Kind))
+			}
+			log.Info(fmt.Sprintf("Could not find %s, requeuing", cluster.Spec.ControlPlaneRef.Kind))
+			return ctrl.Result{RequeueAfter: externalReadyWait}, nil
+		}
+		return ctrl.Result{}, err
 	}
+	s.controlPlane = obj
 
-	preReconcileControlPlaneReady := cluster.Status.ControlPlaneReady
 	// Determine if the control plane provider is ready.
-	ready, err := external.IsReady(controlPlaneConfig)
+	ready, err := external.IsReady(s.controlPlane)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
-	cluster.Status.ControlPlaneReady = ready
-	// Only record the event if the status has changed
-	if preReconcileControlPlaneReady != cluster.Status.ControlPlaneReady {
-		r.recorder.Eventf(cluster, corev1.EventTypeNormal, "ControlPlaneReady", "Cluster %s ControlPlaneReady is now %t", cluster.Name, cluster.Status.ControlPlaneReady)
-	}
 
 	// Report a summary of current status of the control plane object defined for this cluster.
 	conditions.SetMirror(cluster, clusterv1.ControlPlaneReadyCondition,
```

Comment:
I think we also should have an if/else for fallBack like we have in machine_controller_phases.go